perfkitbenchmarker/providers/profitbricks/profitbricks_virtual_machine.py

```diff
@@ -0,0 +1,368 @@
+
+  Attributes:
+    cores: int. Number of CPU cores for a custom VM.
+    ram: int. Amount of RAM in MBs for a custom VM.
+  """
+
+  @classmethod
+  def _GetOptionDecoderConstructions(cls):
+    """Gets decoder classes and constructor args for each configurable option.
+
+    Returns:
+      dict. Maps option name string to a (ConfigOptionDecoder class, dict) pair.
+          The pair specifies a decoder class and its __init__() keyword
+          arguments to construct in order to decode the named option.
+    """
+    result = super(CustomMachineTypeSpec, cls)._GetOptionDecoderConstructions()
+    result.update({'cores': (option_decoders.IntDecoder, {'min': 1}),
+                   'ram': (option_decoders.IntDecoder, {'min': 1024})})
+    return result
+
+
+class MachineTypeDecoder(option_decoders.TypeVerifier):
+  """Decodes the machine_type option of a ProfitBricks VM config."""
+
+  def __init__(self, **kwargs):
+    super(MachineTypeDecoder, self).__init__((basestring, dict), **kwargs)
+
+  def Decode(self, value, component_full_name, flag_values):
+    """Decodes the machine_type option of a ProfitBricks VM config.
+
+    Args:
+      value: Either a string name of a PB machine type or a dict containing
+          'cpu' and 'memory' keys describing a custom VM.
+      component_full_name: string. Fully qualified name of the configurable
+          component containing the config option.
+      flag_values: flags.FlagValues. Runtime flag values to be propagated to
+          BaseSpec constructors.
+
+    Returns:
+      If value is a string, returns it unmodified. Otherwise, returns the
+      decoded CustomMachineTypeSpec.
+
+    Raises:
+      errors.Config.InvalidValue upon invalid input value.
+    """
+    super(MachineTypeDecoder, self).Decode(value, component_full_name,
+                                           flag_values)
+    if isinstance(value, basestring):
+      return value
+    return CustomMachineTypeSpec(self._GetOptionFullName(component_full_name),
+                                 flag_values=flag_values, **value)
+
+
+class ProfitBricksVmSpec(virtual_machine.BaseVmSpec):
+  """Object containing the information needed to create a
+  ProfitBricksVirtualMachine.
+
+  Attributes:
+    ram: None or int. RAM value in MB for custom ProfitBricks VM.
+    cores: None or int. CPU cores value for custom ProfitBricks VM.
+  """
+
+  CLOUD = providers.PROFITBRICKS
+
+  def __init__(self, *args, **kwargs):
+    super(ProfitBricksVmSpec, self).__init__(*args, **kwargs)
+    if isinstance(self.machine_type, CustomMachineTypeSpec):
+      logging.info('Using custom hardware configuration from config.')
+      self.cores = self.machine_type.cores
+      self.ram = self.machine_type.ram
+      self.machine_type = 'Custom (RAM: {}, Cores: {})'.format(self.ram,
+                                                               self.cores)
+    else:
+      try:
+          hardware = ast.literal_eval(self.machine_type)
```

Comment:
Strings seem to work fine now.  I've tried passing "Small" through the command line flag, in the `default_config_constants.yaml`, and in a user_config file using the `--benchmark_config_file` flag, and all three methods work perfectly.

I'm able to pass custom values through the command line flag, and `default_config_constants.yaml`, but it fails with the following error when I try to use custom values in a user_config file.
```
$ ./pkb.py --cloud=ProfitBricks --benchmarks=ping --benchmark_config_file=user_config.yaml
2016-08-22 14:56:47,122 3a7b16c1 MainThread INFO     Verbose logging to: /var/folders/8f/2q6dw27d2tsbhzf41vmt3gv80000gn/T/perfkitbenchmarker/runs/3a7b16c1/pkb.log
2016-08-22 14:56:47,122 3a7b16c1 MainThread INFO     PerfKitBenchmarker version: v1.4.0-106-g7f829f7
2016-08-22 14:56:47,122 3a7b16c1 MainThread INFO     Flag values:
--benchmark_config_file=user_config.yaml
--cloud=ProfitBricks
--benchmarks=ping
2016-08-22 14:56:47,558 3a7b16c1 MainThread WARNING  The key "flags" was not in the default config, but was in user overrides. This may indicate a typo.
2016-08-22 14:56:47,655 3a7b16c1 MainThread INFO     flag: {'cores': 2, 'ram': 4096}
2016-08-22 14:56:47,655 3a7b16c1 MainThread INFO     config: {'cores': 2, 'ram': 4096}
Traceback (most recent call last):
  File "./pkb.py", line 21, in <module>
    sys.exit(Main())
  File "/Users/tyler-macbook/Projects/PerfKitBenchmarker/perfkitbenchmarker/pkb.py", line 662, in Main
    return RunBenchmarks()
  File "/Users/tyler-macbook/Projects/PerfKitBenchmarker/perfkitbenchmarker/pkb.py", line 582, in RunBenchmarks
    benchmark_run_list = _CreateBenchmarkRunList()
  File "/Users/tyler-macbook/Projects/PerfKitBenchmarker/perfkitbenchmarker/pkb.py", line 289, in _CreateBenchmarkRunList
    flag_values=FLAGS, **config_dict)
  File "/Users/tyler-macbook/Projects/PerfKitBenchmarker/perfkitbenchmarker/configs/benchmark_config_spec.py", line 408, in __init__
    super(BenchmarkConfigSpec, self).__init__(component_full_name, **kwargs)
  File "/Users/tyler-macbook/Projects/PerfKitBenchmarker/perfkitbenchmarker/configs/spec.py", line 81, in __init__
    flag_values)
  File "/Users/tyler-macbook/Projects/PerfKitBenchmarker/perfkitbenchmarker/configs/benchmark_config_spec.py", line 464, in _DecodeAndInit
    component_full_name, config, decoders, flag_values)
  File "/Users/tyler-macbook/Projects/PerfKitBenchmarker/perfkitbenchmarker/configs/spec.py", line 145, in _DecodeAndInit
    value = decoder.Decode(config[option], component_full_name, flag_values)
  File "/Users/tyler-macbook/Projects/PerfKitBenchmarker/perfkitbenchmarker/configs/benchmark_config_spec.py", line 353, in Decode
    flag_values=flag_values, **vm_group_config)
  File "/Users/tyler-macbook/Projects/PerfKitBenchmarker/perfkitbenchmarker/configs/benchmark_config_spec.py", line 275, in __init__
    flag_values=flag_values, **vm_config)
  File "/Users/tyler-macbook/Projects/PerfKitBenchmarker/perfkitbenchmarker/providers/profitbricks/profitbricks_virtual_machine.py", line 108, in __init__
    super(ProfitBricksVmSpec, self).__init__(*args, **kwargs)
  File "/Users/tyler-macbook/Projects/PerfKitBenchmarker/perfkitbenchmarker/configs/spec.py", line 69, in __init__
    self._ApplyFlags(kwargs, flag_values)
  File "/Users/tyler-macbook/Projects/PerfKitBenchmarker/perfkitbenchmarker/providers/profitbricks/profitbricks_virtual_machine.py", line 134, in _ApplyFlags
    config_values['machine_type'] = yaml.load(flag_values.machine_type)
  File "/usr/local/lib/python2.7/site-packages/yaml/__init__.py", line 69, in load
    loader = Loader(stream)
  File "/usr/local/lib/python2.7/site-packages/yaml/loader.py", line 34, in __init__
    Reader.__init__(self, stream)
  File "/usr/local/lib/python2.7/site-packages/yaml/reader.py", line 85, in __init__
    self.determine_encoding()
  File "/usr/local/lib/python2.7/site-packages/yaml/reader.py", line 124, in determine_encoding
    self.update_raw()
  File "/usr/local/lib/python2.7/site-packages/yaml/reader.py", line 178, in update_raw
    data = self.stream.read(size)
AttributeError: 'dict' object has no attribute 'read'
```
My guess is that this is likely an error in my user_config file.  I tried to keep it very simple and just update the machine_type flag, like so:
```
ping:
  flags:
    machine_type:
      cores: 2
      ram: 4096
```
For some reason, `config_values['machine_type']` and `flag_values['machine_type']` both already have the same value before the `yaml.load()` method executes and dies.  
```
2016-08-22 14:56:47,655 3a7b16c1 MainThread INFO     flag: {'cores': 2, 'ram': 4096}
2016-08-22 14:56:47,655 3a7b16c1 MainThread INFO     config: {'cores': 2, 'ram': 4096}
```

Anyways, I've made all of the changes outlined in your other comments.  I'll push everything up to GH so that you have eyes on it.  I'll keep poking away at this user_config issue.